### PR TITLE
feat: draft of ffetch, first operators

### PR DESCRIPTION
**Pull request by buuhuu**
### Usage

**async generator**

```
const tenTitles = ffetch('/path-to-index.json')
  .limit(10)
  .filter(({ path }) => path.indexOf('/foobar') === 0)
  .map(({ title }) => title);

for await (const title of tenTitles) console.log(title); 
```

**promise**

```
const tenTitles = await ffetch('/path-to-index.json')
  .filter(({ path }) => path.indexOf('/foobar') === 0)
  .slice(20, 30)
  .map(({ title }) => title)
  .all()


for (const title of tenTitles) console.log(title); 
```

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+  env: {
+    browser: true,
+    es2021: true,
+  },
+  extends: 'airbnb-base',
```
Comment (trieloff) [inline .eslintrc.cjs]: Good start, we can re-base this to the `helix-eslint-config` which is an extension of `airbnb-base` later.

File: package.json [hunk 2]
```diff
@@ -15,6 +16,11 @@
   "license": "Apache-2.0",
   "devDependencies": {
```
Comment (trieloff) [inline package.json:17]: We should add a test to make sure that `package.json/dependencies` is always empty :)

File: src/ffetch.js [hunk 3]
```diff
@@ -9,4 +8,75 @@
  * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  * OF ANY KIND, either express or implied. See the License for the specific language
  * governing permissions and limitations under the License.
- */
\ No newline at end of file
+ */
+
+async function* request({
+  url, offset, chunks, skip, limit, filter, map, fetch,
+}) {
```
Comment (trieloff) [inline src/ffetch.js]: This will likely need a refactoring, as it will become unwieldy when more helper functions are added.

**Review by trieloff — APPROVED** [review]
Good start.

File: .eslintrc.cjs (hunk 1 above)
Comment (buuhuu) [reply .eslintrc.cjs]: done

File: package.json (hunk 2 above)
Comment (buuhuu) [reply package.json:17]: done

File: src/ffetch.js [hunk 4]
```diff
@@ -9,4 +8,128 @@
...
+
+function filter(upstream, fn) {
+  return map(upstream, (entry) => (fn(entry) ? entry : null));
+}
+
+function slice(upstream, from, to) {
+  return limit(skip(upstream, from), to - from);
+}
+
+function follow(upstream, name, context) {
```
Comment (duynguyen) [inline src/ffetch.js]: It's unclear to me what the use case of `follow` is?
to pre-fetch the reference content on index page?

File: src/ffetch.js [hunk 5]
```diff
@@ -9,4 +8,128 @@
...
+async function* request(url, context) {
+  const { chunks, fetch } = context;
+  for (let offset = 0, total = Infinity; offset < total; offset += chunks) {
+    const resp = await fetch(`${url}?offset=${offset}&limit=${chunks}`);
+    if (resp.ok) {
+      const json = await resp.json();
+      total = json.total;
+      for (const entry of json.data) yield entry;
+    } else {
+      return;
```
Comment (trieloff) [inline src/ffetch.js:22]: I can see the `.retry(3)` feature coming already :)

File: src/ffetch.js (hunk 4 above)
Comment (trieloff) [reply src/ffetch.js]: It will pre-fetch it, but it will also parse it, so you can use content from the referenced document in your overview page.

**Review by trieloff — APPROVED** [review]

File: src/ffetch.js (hunk 5 above)
Comment (buuhuu) [reply src/ffetch.js:22]: we can add some error handling when we see the need for it.

Comment (duynguyen) [reply src/ffetch.js:22]: in addition, we can retry with [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff). Not a blocker though, can be added based on demand.

**Review by duynguyen — APPROVED** [review]
I think this is nice for a first release. We can always make more improvements as discussed in the comments, but good to have users for it at first.

**Pull request merged**